Assets/Mirror/Runtime/Authenticator.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections;
+using UnityEngine;
+using UnityEngine.Events;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Unity Event for the NetworkConnection
+    /// </summary>
+    [Serializable] public class UnityEventNetworkConnection : UnityEvent<NetworkConnection> { }
+
+    /// <summary>
+    /// Base class for implementing component-based authentication during the Connect phase
+    /// </summary>
+    [HelpURL("https://mirror-networking.com/xmldocs/articles/Concepts/Authentication.html")]
+    public abstract class Authenticator : MonoBehaviour
+    {
+        [Header("Event Listeners (optional)")]
+
+        /// <summary>
+        /// Notify subscribers on the server when a client is authenticated
+        /// </summary>
+        [Tooltip("Mirror has an internal subscriber to this event. You can add your own here.")]
+        public UnityEventNetworkConnection OnServerAuthenticated = new UnityEventNetworkConnection();
+
+        /// <summary>
+        /// Notify subscribers on the client when the client is authenticated
+        /// </summary>
+        [Tooltip("Mirror has an internal subscriber to this event. You can add your own here.")]
+        public UnityEventNetworkConnection OnClientAuthenticated = new UnityEventNetworkConnection();
+
+        #region server
+
+        /// <summary>
+        /// Called on server from StartServer to initialize the Authenticator
+        /// <para>Server message handlers should be registered in this method.</para>
+        /// </summary>
+        public abstract void OnStartServer();
+
+        /// <summary>
+        /// Called on client from StartClient to initialize the Authenticator
+        /// <para>Client message handlers should be registered in this method.</para>
+        /// </summary>
+        public abstract void OnStartClient();
+
+        // This will get more code in the near future
+        internal void OnServerAuthenticateInternal(NetworkConnection conn)
```

Comment:
OnServerAuthenticate + OnClientAuthenticate should be enough, we don't need the internal ones here.
https://en.wikipedia.org/wiki/KISS_principle

Reply:
Keeping them for when we put the timeout back in.